File: quest-client-unity/ai_1/Assets/Scripts/Networking/Messages.cs
using System;

namespace QuestAssistant
{
    // 与 shared/schemas + docs/protocol.md 对齐的线上消息结构。
    // 用 Unity JsonUtility 序列化，字段名必须与后端逐字一致（snake_case）。

    [Serializable]
    public class PointingEvent
    {
        public string type = "pointing_event";
        public string session_id;
        public long timestamp_ms;
        public string target_id;
        public string target_label;
        public float[] ray_origin;
        public float[] ray_direction;
    }

    [Serializable]
    public class AssistantResponse
    {
        public string type;
        public string target_id;
        public string speech;
        public string display_title;
        public string display_subtitle;
        public string status;
        public string avatar_intent;
        public string tts_mode;
    }

    // 仅用于先读出 type 字段，判断这条消息是 assistant_response 还是 error。
    [Serializable]
    public class MessageEnvelope
    {
        public string type;
        public string message; // error 消息体
    }
}

File: backend/app/schemas/messages.py
"""与 shared/schemas 对齐的 Pydantic 模型。

协议契约见 ../../../shared/schemas/ 与 docs/protocol.md。
camera_* 字段是 Quest 3 预留的可选扩展，V0 不使用。
"""
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class PointingEvent(BaseModel):
    type: Literal["pointing_event"] = "pointing_event"
    session_id: str
    timestamp_ms: int
    target_id: str
    target_label: Optional[str] = None
    ray_origin: Optional[list[float]] = None
    ray_direction: Optional[list[float]] = None

    # —— Quest 3 预留（可选扩展，V0 不用）——
    camera_frame_id: Optional[str] = None
    camera_intrinsics: Optional[dict[str, Any]] = None
    camera_pose: Optional[dict[str, Any]] = None
    image_width: Optional[int] = None
    image_height: Optional[int] = None


class AssistantResponse(BaseModel):
    type: Literal["assistant_response"] = "assistant_response"
    target_id: str
    speech: str
    display_title: Optional[str] = None
    display_subtitle: Optional[str] = None
    avatar_intent: Literal["none", "look_and_point", "look"] = "look_and_point"
    tts_mode: Literal["local", "server"] = "local"


class ErrorResponse(BaseModel):
    type: Literal["error"] = "error"
    message: str
    target_id: Optional[str] = None
